db/change_listener.go

```diff
@@ -212,10 +221,49 @@ func (listener *changeListener) Notify(ctx context.Context, keys channels.Set) {
 	}
 	base.DebugfCtx(ctx, base.KeyChanges, "Notifying that %q changed (keys=%q) count=%d",
 		base.MD(listener.bucketName), base.UD(keys), listener.counter)
-	listener.tapNotifier.Broadcast()
 	listener.tapNotifier.L.Unlock()
 }
 
+func (listener *changeListener) BroadcastChanges(ctx context.Context) {
+	ticker := time.NewTicker(DefaultBroadcastChangesTime)
+	// boolean to indicate whether ticker is using the default value, this is needed so we don't call reset on ticker
+	// for a value it already has
+	defaultTickerValue := true
+	terminator := listener.terminator // take copy of this to avoid race when restarting change listener in tests
+	go func() {
+		var currCount uint64
+		for {
+			select {
+			case <-terminator:
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				// if the counter has changed, notify waiting clients
+				listener.tapNotifier.L.Lock()
+				if listener.counter > currCount {
+					listener.tapNotifier.Broadcast()
+					currCount = listener.counter
+				}
+				listener.tapNotifier.L.Unlock()
+
+				// check if we need to reset ticker value based on skipped sequence presence
+				shouldUseSkippedValue := listener.SkippedSequenceBroadcast.Load()
+				if shouldUseSkippedValue && defaultTickerValue {
+					// change the ticker from default value to use skipped sequence value
+					base.DebugfCtx(ctx, base.KeyChanges, "Changing broadcast changes interval to the skipped sequence value for %q", base.MD(listener.bucketName))
+					ticker.Reset(SkippedSequenceBroadcastChangesTime)
+					defaultTickerValue = false // update local bool to indicate ticker value is no longer using the default value
+				} else if !shouldUseSkippedValue && !defaultTickerValue {
+					// change the ticker from skipped sequence value to default value
+					base.DebugfCtx(ctx, base.KeyChanges, "Changing broadcast changes interval to the default value for %q", base.MD(listener.bucketName))
+					ticker.Reset(DefaultBroadcastChangesTime)
+					defaultTickerValue = true // update local bool to indicate ticker value is using the default value
+				}
```

Comment:
If we switch both bools to mean the same thing (e.g. `slowMode`), we can simplify the logic when changing to something like this, because anything non-matching means it needs to change:

where `tickerValForBroadcastSpeed` is just a function that returns a new ticker duration for the given `true`/`false` value of `slowMode`

```suggestion
				// change the ticker duration if required
				newSlowMode := listener.broadcastSlowMode.Load()
				if newSlowMode != broadcastSlowMode {
					newDuration := tickerValForBroadcastSpeed(slowMode)
					base.DebugfCtx(ctx, base.KeyChanges, "Updating broadcast changes interval for %q to %v", base.MD(listener.bucketName), newDuration)
					broadcastSlowMode = newSlowMode
					ticker.Reset(newDuration)
				}
```